File: mcp/src/tools/gsc-inspect.ts
import { z } from 'zod';

// ============================================================================
// GSC URL Inspect Tool
// ============================================================================

/**
 * Input schema for gsc_inspect_url tool
 */
export const gscInspectUrlInputSchema = z.object({
  /** Site URL: domain property (sc-domain:example.com) or URL prefix (https://example.com/) */
  site: z.string().min(1, 'Site URL is required'),
  /** URL to inspect (must be http or https) */
  url: z.string().min(1, 'URL to inspect is required'),
});

export type GscInspectUrlInput = z.infer<typeof gscInspectUrlInputSchema>;

/**
 * Indexing issue from URL inspection
 */
export interface IndexingIssue {
  severity: 'ERROR' | 'WARNING';
  issue_type: string;
  message: string;
}

/**
 * Quota status for URL inspection API
 */
export interface InspectQuotaStatus {
  used: number;
  limit: number;
  remaining: number;
  date?: string;
  warning?: string;
}

/**
 * URL inspection result output structure
 */
export interface InspectUrlOutput {
  success: boolean;
  operation: 'inspect_url';
  url?: string;
  verdict?: 'PASS' | 'PARTIAL' | 'FAIL' | 'NEUTRAL';
  coverage_state?: string;
  last_crawl?: string;
  google_canonical?: string;
  user_canonical?: string;
  indexing_allowed?: boolean;
  robots_blocked?: boolean;
  mobile_usability?: 'PASS' | 'FAIL';
  mobile_issues?: string[];
  rich_results_status?: 'PASS' | 'FAIL';
  rich_results_issues?: string[];
  issues: IndexingIssue[];
  quota?: InspectQuotaStatus;
  error?: string;
}

/**
 * Build CLI arguments for URL inspection
 */
export function buildInspectUrlArgs(input: GscInspectUrlInput): string[] {
  return ['gsc', 'inspect', 'url', '--site', input.site, '--url', input.url];
}

/**
 * Parse verdict from index status line
 */
function parseVerdict(output: string): 'PASS' | 'PARTIAL' | 'FAIL' | 'NEUTRAL' | undefined {
  // Check for different verdict indicators
  if (output.includes('Indexed (PASS)') || /✓\s*Indexed\s*\(PASS\)/.test(output)) {
    return 'PASS';
  }
  if (output.includes('Partially Indexed (PARTIAL)') || /⚠\s*Partially Indexed\s*\(PARTIAL\)/.test(output)) {
    return 'PARTIAL';
  }
  if (output.includes('Not Indexed (FAIL)') || /✗\s*Not Indexed\s*\(FAIL\)/.test(output)) {
    return 'FAIL';
  }
  if (output.includes('(NEUTRAL)')) {
    return 'NEUTRAL';
  }

  // Direct status pattern matching
  const statusMatch = output.match(/Status:\s*(PASS|PARTIAL|FAIL|NEUTRAL)/);
  if (statusMatch) {
    return statusMatch[1] as 'PASS' | 'PARTIAL' | 'FAIL' | 'NEUTRAL';
  }

  return undefined;
}

/**
 * Parse mobile usability verdict
 */
function parseMobileUsability(output: string): 'PASS' | 'FAIL' | undefined {
  // Check for "Not Mobile Usable" first (more specific pattern)
  if (output.includes('Not Mobile Usable') || /✗\s*Not Mobile Usable/.test(output)) {
    return 'FAIL';
  }
  // Then check for positive "Mobile Usable" (less specific pattern)
  if (output.includes('Mobile Usable') || /✓\s*Mobile Usable/.test(output)) {
    return 'PASS';
  }
  return undefined;
}

/**
 * Parse issues from the issues table
 */
function parseIssues(output: string): IndexingIssue[] {
  const issues: IndexingIssue[] = [];

  // Match table rows: | SEVERITY | ISSUE TYPE | MESSAGE |
  const tableRowRegex = /\|\s*(ERROR|WARNING)\s*\|\s*([^|]+)\s*\|\s*([^|]+)\s*\|/g;

  let match;
  while ((match = tableRowRegex.exec(output)) !== null) {
    issues.push({
      severity: match[1].trim() as 'ERROR' | 'WARNING',
      issue_type: match[2].trim(),
      message: match[3].trim(),
    });
  }

  return issues;
}

/**
 * Parse mobile issues from output
 */
function parseMobileIssues(output: string): string[] {
  const issues: string[] = [];

  // Look for Mobile Issues section
  const mobileSection = output.match(/Mobile Issues:\s*([\s\S]*?)(?=\n\n|Rich Results:|Issues Found:|$)/);
  if (mobileSection) {
    const issueLines = mobileSection[1].match(/-\s*(.+)/g);
    if (issueLines) {
      for (const line of issueLines) {
        const issueMatch = line.match(/-\s*(.+)/);
        if (issueMatch) {
          issues.push(issueMatch[1].trim());
        }
      }
    }
  }

  return issues;
}

/**
 * Parse rich results issues from output
 */
function parseRichResultsIssues(output: string): string[] {
  const issues: string[] = [];

  // Look for Rich Results Issues section
  const richSection = output.match(/Rich Results Issues:\s*([\s\S]*?)(?=\n\n|Issues Found:|$)/);
  if (richSection) {
    const issueLines = richSection[1].match(/-\s*(.+)/g);
    if (issueLines) {
      for (const line of issueLines) {
        const issueMatch = line.match(/-\s*(.+)/);
        if (issueMatch) {
          issues.push(issueMatch[1].trim());
        }
      }
    }
  }

  return issues;
}

/**
 * Parse quota status from output
 */
function parseQuotaStatus(output: string): InspectQuotaStatus | undefined {
  // Pattern: Inspections: X / Y (Z% used, N remaining)
  const quotaMatch = output.match(/Inspections:\s*(\d+)\s*\/\s*(\d+)\s*\([^)]+,\s*(\d+)\s*remaining\)/);
  if (quotaMatch) {
    const used = parseInt(quotaMatch[1], 10);
    const limit = parseInt(quotaMatch[2], 10);
    const remaining = parseInt(quotaMatch[3], 10);

    const quota: InspectQuotaStatus = {
      used,
      limit,
      remaining,
    };

    // Extract date if present
    const dateMatch = output.match(/Date:\s*(\d{4}-\d{2}-\d{2})/);
    if (dateMatch) {
      quota.date = dateMatch[1];
    }

    // Check for warnings
    if (output.includes('CRITICAL: Approaching daily limit')) {
      quota.warning = 'CRITICAL: Approaching daily limit';
    } else if (output.includes('WARNING:')) {
      const warningMatch = output.match(/WARNING:\s*([^\n]+)/);
      if (warningMatch) {
        quota.warning = warningMatch[1].trim();
      }
    }

    return quota;
  }

  return undefined;
}

/**
 * Parse CLI output for URL inspection
 */
export function parseInspectUrlOutput(output: string): InspectUrlOutput {
  const result: InspectUrlOutput = {
    success: true,
    operation: 'inspect_url',
    issues: [],
  };

  // Check for errors
  if (output.includes('Failed to inspect URL:')) {
    const errorMatch = output.match(/Failed to inspect URL:\s*(.+)/);
    result.success = false;
    result.error = errorMatch ? errorMatch[1].trim() : 'Unknown error';
    return result;
  }

  if (output.includes('Failed to create GSC client:')) {
    const errorMatch = output.match(/Failed to create GSC client:\s*(.+)/);
    result.success = false;
    result.error = errorMatch ? errorMatch[1].trim() : 'Failed to create GSC client';
    return result;
  }

  // Check for quota exceeded
  if (output.includes('daily quota critical threshold reached')) {
    result.success = false;
    result.error = 'Daily quota critical threshold reached. Please wait until tomorrow to continue.';
    return result;
  }

  // Extract URL
  const urlMatch = output.match(/URL:\s*(https?:\/\/[^\s\n]+)/);
  if (urlMatch) {
    result.url = urlMatch[1].trim();
  }

  // Extract verdict/index status
  result.verdict = parseVerdict(output);

  // Extract coverage state
  const coverageMatch = output.match(/Coverage:\s*(.+)/);
  if (coverageMatch) {
    result.coverage_state = coverageMatch[1].trim();
  }

  // Extract last crawl time
  const crawlMatch = output.match(/Last Crawl:\s*(.+)/);
  if (crawlMatch) {
    result.last_crawl = crawlMatch[1].trim();
  }

  // Extract Google canonical URL
  const googleCanonicalMatch = output.match(/Google Canonical:\s*(https?:\/\/[^\s\n]+)/);
  if (googleCanonicalMatch) {
    result.google_canonical = googleCanonicalMatch[1].trim();
  }

  // Extract user canonical URL
  const userCanonicalMatch = output.match(/User Canonical:\s*(https?:\/\/[^\s\n]+)/);
  if (userCanonicalMatch) {
    result.user_canonical = userCanonicalMatch[1].trim();
  }

  // Extract indexing allowed status
  if (output.includes('Indexing Allowed') || /✓\s*Indexing Allowed/.test(output)) {
    result.indexing_allowed = true;
  } else if (output.includes('Indexing Not Allowed') || /✗\s*Indexing Not Allowed/.test(output)) {
    result.indexing_allowed = false;
  }

  // Extract robots blocked status
  if (output.includes('Blocked by robots.txt') || /✗\s*Blocked by robots\.txt/.test(output)) {
    result.robots_blocked = true;
  } else {
    result.robots_blocked = false;
  }

  // Extract mobile usability
  result.mobile_usability = parseMobileUsability(output);

  // Extract mobile issues
  const mobileIssues = parseMobileIssues(output);
  if (mobileIssues.length > 0) {
    result.mobile_issues = mobileIssues;
  }

  // Extract rich results status
  if (output.includes('Rich Results:')) {
    if (output.includes('Valid (PASS)') || /✓\s*Valid\s*\(PASS\)/.test(output)) {
      result.rich_results_status = 'PASS';
    } else if (output.includes('Invalid (FAIL)') || /✗\s*Invalid\s*\(FAIL\)/.test(output)) {
      result.rich_results_status = 'FAIL';
    }
  }

  // Extract rich results issues
  const richIssues = parseRichResultsIssues(output);
  if (richIssues.length > 0) {
    result.rich_results_issues = richIssues;
  }

  // Parse indexing issues from table
  result.issues = parseIssues(output);

  // Parse quota status
  result.quota = parseQuotaStatus(output);

  return result;
}

/**
 * MCP Tool definition for gsc_inspect_url
 */
export const gscInspectUrlTool = {
  name: 'gsc_inspect_url',
  description:
    'Inspect URL indexing status in Google Search Console. Shows whether a URL is indexed, coverage state, crawl info, mobile usability, rich results, and any indexing issues. Rate limits: 2,000 inspections/day, 600/minute per property.',
  inputSchema: {
    type: 'object',
    properties: {
      site: {
        type: 'string',
        description:
          'Site URL: domain property (sc-domain:example.com - RECOMMENDED) or URL prefix (https://example.com/)',
      },
      url: {
        type: 'string',
        description: 'URL to inspect (e.g., https://example.com/page)',
      },
    },
    required: ['site', 'url'],
  },
};

// ============================================================================
// Exports
// ============================================================================

export const gscInspectTools = [gscInspectUrlTool] as const;

File: cmd/gsc_inspect.go
package cmd

import (
	"fmt"
	"os"
	"time"

	"github.com/fatih/color"
	"github.com/spf13/cobra"

	"github.com/garbarok/ga4-manager/internal/gsc"
	"github.com/garbarok/ga4-manager/internal/render"
)

var (
	gscInspectURL      string
	gscRichResultsOnly bool
)

var gscInspectCmd = &cobra.Command{
	Use:   "inspect",
	Short: "Inspect URL indexing status in Google Search Console",
	Long: `Check if URLs are indexed by Google and view coverage issues.

The URL Inspection API provides detailed information about:
  - Index status (indexed, excluded, error)
  - Coverage state (submitted and indexed, not found, blocked, etc.)
  - Mobile usability issues
  - Rich results validation
  - Robots.txt blocking
  - Last crawl time

Property Types:
  - Domain property: sc-domain:example.com (RECOMMENDED - covers all subdomains and protocols)
  - URL prefix: https://example.com/ (less flexible, exact URL match, must end with /)

Note: Domain properties are more reliable and flexible. Use them unless you specifically
need URL prefix properties.

Rate Limits:
  - 2,000 URL inspections per day
  - 600 inspections per minute per property

Examples:
  # Inspect a single URL (domain property - RECOMMENDED)
  ga4 gsc inspect url --site sc-domain:example.com --url https://example.com/page

  # Inspect homepage
  ga4 gsc inspect url --site sc-domain:example.com --url https://example.com/

  # Inspect a blog post
  ga4 gsc inspect url --site sc-domain:example.com --url https://example.com/blog/post

  # Using URL prefix property (alternative, less flexible)
  ga4 gsc inspect url --site https://example.com/ --url https://example.com/page`,
}

var gscInspectURLCmd = &cobra.Command{
	Use:   "url",
	Short: "Inspect a single URL",
	Long:  "Check the indexing status of a specific URL and view any coverage or mobile usability issues.",
	RunE:  runGSCInspectURL,
}

func init() {
	gscCmd.AddCommand(gscInspectCmd)
	gscInspectCmd.AddCommand(gscInspectURLCmd)

	// Site URL flag (required, inherited from parent)
	gscInspectCmd.PersistentFlags().StringVarP(&gscSiteURL, "site", "s", "", "Site URL: domain property (sc-domain:example.com) or URL prefix (https://example.com/)")
	_ = gscInspectCmd.MarkPersistentFlagRequired("site")

	// URL flag (required for url command)
	gscInspectURLCmd.Flags().StringVarP(&gscInspectURL, "url", "u", "", "URL to inspect (e.g., https://example.com/page)")
	_ = gscInspectURLCmd.MarkFlagRequired("url")

	// Rich results only flag (optional)
	gscInspectURLCmd.Flags().BoolVarP(&gscRichResultsOnly, "rich-results-only", "r", false, "Show only rich results information")
}

func runGSCInspectURL(cmd *cobra.Command, args []string) error {
	// Create GSC client
	client, err := gsc.NewClient()
	if err != nil {
		color.Red("✗ Failed to create GSC client: %v", err)
		return err
	}
	defer func() { _ = client.Close() }()

	// Display progress
	color.Cyan("🔍 Inspecting URL: %s", gscInspectURL)
	fmt.Println()

	// Call API
	result, err := client.InspectURL(gscSiteURL, gscInspectURL)
	if err != nil {
		color.Red("✗ Failed to inspect URL: %v", err)
		return err
	}

	// Display detailed results
	if err := displayInspectionResult(result, gscRichResultsOnly); err != nil {
		return err
	}

	// Display quota status (skip if rich-results-only mode)
	if !gscRichResultsOnly {
		displayInspectQuotaStatus(client)
	}

	return nil
}

func displayInspectionResult(result *gsc.URLInspectionResult, richResultsOnly bool) error {
	// Header
	if richResultsOnly {
		color.Cyan("═══ Rich Results Validation ═══")
	} else {
		color.Cyan("═══ URL Inspection Results ═══")
	}
	fmt.Println()

	// URL
	fmt.Printf("URL: %s\n", result.URL)
	fmt.Println()

	// If rich-results-only mode, skip to rich results section
	if richResultsOnly {
		displayRichResults(result)
		return nil
	}

	// Index Status
	color.Cyan("Index Status:")
	status := result.IndexStatus
	switch status {
	case "PASS":
		color.Green("  ✓ Indexed (%s)", status)
	case "PARTIAL":
		color.Yellow("  ⚠ Partially Indexed (%s)", status)
	case "FAIL":
		color.Red("  ✗ Not Indexed (%s)", status)
	default:
		fmt.Printf("  Status: %s\n", status)
	}

	// Coverage State
	if result.CoverageState != "" {
		fmt.Printf("  Coverage: %s\n", result.CoverageState)
	}
	fmt.Println()

	// Crawl Information
	if result.LastCrawlTime != "" {
		color.Cyan("Crawl Information:")
		// Parse and format time
		if t, err := time.Parse(time.RFC3339, result.LastCrawlTime); err == nil {
			fmt.Printf("  Last Crawl: %s\n", t.Format("2006-01-02 15:04:05 MST"))
		} else {
			fmt.Printf("  Last Crawl: %s\n", result.LastCrawlTime)
		}
		fmt.Println()
	}

	// Canonical URLs
	if result.GoogleCanonical != "" || result.UserCanonical != "" {
		color.Cyan("Canonical URLs:")
		if result.GoogleCanonical != "" {
			fmt.Printf("  Google Canonical: %s\n", result.GoogleCanonical)
		}
		if result.UserCanonical != "" {
			fmt.Printf("  User Canonical: %s\n", result.UserCanonical)
		}
		fmt.Println()
	}

	// Indexing Allowed
	color.Cyan("Indexing Status:")
	if result.IndexingAllowed {
		color.Green("  ✓ Indexing Allowed")
	} else {
		color.Red("  ✗ Indexing Not Allowed")
	}

	if result.RobotsBlocked {
		color.Red("  ✗ Blocked by robots.txt")
	}
	fmt.Println()

	// Mobile Usability
	color.Cyan("Mobile Usability:")
	switch {
	case !result.MobileUsabilityChecked:
		// Google deprecated the Mobile Usability report/API field (Dec 2023);
		// no verdict is returned, so this is "unknown", not a failure.
		color.HiBlack("  – Not reported (Google deprecated this signal in Dec 2023)")
	case result.MobileUsable:
		color.Green("  ✓ Mobile Usable")
	default:
		color.Red("  ✗ Not Mobile Usable")
	}

	if len(result.MobileIssues) > 0 {
		color.Yellow("  Mobile Issues:")
		for _, issue := range result.MobileIssues {
			fmt.Printf("    - %s\n", issue)
		}
	}
	fmt.Println()

	// Rich Results
	displayRichResults(result)

	// Indexing Issues Summary
	if len(result.IndexingIssues) > 0 {
		color.Cyan("Issues Found:")
		if err := render.Render(os.Stdout, render.FormatTable, inspectIssuesColumns(), result.IndexingIssues, inspectIssuesTableRow); err != nil {
			return fmt.Errorf("failed to render issues table: %w", err)
		}
		fmt.Println()
	} else {
		color.Green("✓ No issues detected")
		fmt.Println()
	}
	return nil
}

// inspectIssuesColumns / inspectIssuesTableRow project an indexing issue for
// the URL-inspection results table. The severity cell keeps fatih/color
// escape codes so terminal output retains its colour cues — matching the
// previous hand-rolled tablewriter output.
func inspectIssuesColumns() []string {
	return []string{"Severity", "Issue Type", "Message"}
}

func inspectIssuesTableRow(issue gsc.IndexingIssue) []string {
	var severity string
	switch issue.Severity {
	case "ERROR":
		severity = color.RedString("ERROR")
	case "WARNING":
		severity = color.YellowString("WARNING")
	default:
		severity = issue.Severity
	}
	return []string{severity, issue.IssueType, issue.Message}
}

// displayRichResults shows rich results information including types and detected items
func displayRichResults(result *gsc.URLInspectionResult) {
	if result.RichResultsStatus == "" {
		color.Yellow("ℹ No rich results data available")
		fmt.Println()
		return
	}

	color.Cyan("Rich Results:")

	// Display verdict
	switch result.RichResultsStatus {
	case "PASS":
		color.Green("  ✓ Valid (%s)", result.RichResultsStatus)
	case "FAIL":
		color.Red("  ✗ Invalid (%s)", result.RichResultsStatus)
	default:
		fmt.Printf("  Status: %s\n", result.RichResultsStatus)
	}

	// Display detected types
	if len(result.RichResultTypes) > 0 {
		fmt.Printf("  Detected Types: %s\n", formatRichResultTypes(result.RichResultTypes))
	}

	// Display detected items with details
	if len(result.RichResultItems) > 0 {
		color.Cyan("\n  Detected Items:")
		for i, item := range result.RichResultItems {
			fmt.Printf("    %d. %s", i+1, item.Type)
			if item.Name != "" {
				fmt.Printf(" - %s", item.Name)
			}
			fmt.Println()

			// Show item-specific issues
			if len(item.Issues) > 0 {
				color.Yellow("       Issues:")
				for _, issue := range item.Issues {
					fmt.Printf("         - %s\n", issue)
				}
			}
		}
	}

	// Display legacy flat issues list (for backward compatibility)
	if len(result.RichResultsIssues) > 0 && len(result.RichResultItems) == 0 {
		color.Yellow("  Rich Results Issues:")
		for _, issue := range result.RichResultsIssues {
			fmt.Printf("    - %s\n", issue)
		}
	}

	fmt.Println()
}

// formatRichResultTypes formats the types array for display
func formatRichResultTypes(types []string) string {
	if len(types) == 0 {
		return "None"
	}
	if len(types) == 1 {
		return types[0]
	}
	// Join with commas
	result := types[0]
	for i := 1; i < len(types); i++ {
		result += ", " + types[i]
	}
	return result
}

func displayInspectQuotaStatus(client *gsc.Client) {
	used, limit, date := client.GetQuotaStatus()
	percentage := float64(used) / float64(limit) * 100

	color.Cyan("═══ Daily Quota Status ═══")
	fmt.Println()

	fmt.Printf("Date: %s\n", date)
	fmt.Printf("Inspections: %d / %d (%.1f%% used, %d remaining)\n",
		used, limit, percentage, limit-used)

	// Show warning if approaching limits
	if percentage >= 95 {
		color.Red("⚠ CRITICAL: Approaching daily limit!")
	} else if percentage >= 75 {
		color.Yellow("⚠ WARNING: %.0f%% of daily quota used", percentage)
	}
	fmt.Println()
}
